### DISCUSSION TITLE
Can you use data from one file in another data file?

### DISCUSSION THREAD

**Discussion by SteGreig** [11ty/eleventy#2647] [Q&A] [answered]
Hi, was wondering if someone might be able to help...

I have a bunch of data files in this structure:

_data/seasons/s2022.json
_data/seasons/s2021.json
_data/seasons/s2020.json
_data/seasons/s2019.json
etc.

Each data file outputs an object containing 2 football players' stats in each competition for that season, so goals, assists, appearances, and a few more.

Ideally, I was thinking in another data file (e.g. _data/all_time.js), I could calculate the totals for each stat. So get the goals total for each competition from each season and then find the sum for each (the all time totals). There are 9 competitions and 7 different stats for each one.

Is there a way to make the data from those season data files available in another global JS data file so I can do all the calculations using JavaScript?

I did try doing these calculations in the template file (.njk) but it doesn't seem possible to do in a nice way. It is technically possible but would be hundreds of lines of the least DRY code imaginable. I thought I had it with the following concept, but it seems you can't update object data in Nunjucks...

```
{% set comps = ['All Club Competitions','League','Champions League','Europa League','Main Domestic Cup','Domestic League Cup','Domestic Super Cup','UEFA Super Cup','Club World Cup'] %}
```

```
{% set stats = {
  'All Club Competitions':[],
  'League':[],
  'Champions League':[],
  'Europa League':[],
  'Main Domestic Cup':[],
  'Domestic League Cup':[],
  'Domestic Super Cup':[],
  'UEFA Super Cup':[],
  'Club World Cup':[]
} %}
```

```
{% for season in seasons %}
  {% for comp in comps %}
      {% set stats[comp] = (stats[comp].push(seasons[season.sname][comp][0][0].goals), stats[comp]) %}
  {% endfor %}
{% endfor %}
```

Can anyone help or suggest an alternative approach? This is my first Eleventy project so there may be a much more sensible way to go about this!

Thanks!

**Comment by pdehaan** [comment, marked as answer]
I'd probably do something like _data/all_time.js or _data/seasons/all_time.js and have that file do a `const s2022 = require("./s2022.json")` (or maybe even something w/ https://www.npmjs.com/package/require-dir) and then do all your calculations in Node (instead of Nunjucks).
Then your final export is just the calculated data which you can loop over.

**Reply by SteGreig** [reply]
Thanks so much - had no idea it could be that simple (bit of a noob to this type of dev), seems extremely obvious now!

Thanks again for your help!

**Reply by SteGreig** [reply]
Sorry @pdehaan, just further to this... I'm really struggling with one issue here.

I can include the .json data files in the all_time.js file and work with that data no problem.

However when I try to require a .js file (which exports an object in the same format), I can export that data as a whole from the all_time.js file but I can't seem to access anything inside of that object.

For example, in all_time.js I can do `module.exports = current` but I can't do `module.exports = current['All Competitions']`. The latter does work in the njk template file however. Why can I not access these properties in the JS file?

Any help would be much appreciated!

**Reply by pdehaan** [reply]
What does the contents of _data/all_time.js look like?

**Reply by pdehaan** [reply]
OK, here's the rough code I had:
### src/_data/seasons/2021.json
```json
{
  "All Competitions": 23,
  "League": 7
}
```

### src/_data/seasons/2022.json
```json
{
  "All Competitions": 19
}
```

### src/_data/seasons/all_time.js
```js
const path = require("node:path");
const fg = require("fast-glob");

const DATA_DIR = path.join(__dirname, "*.json");

const current = fg.sync(DATA_DIR)
  .map(
    filename => require(filename)
  ).reduce((acc, c) => {
    for (const key of Object.keys(c)) {
      if (!acc[key]) {
        acc[key] = 0;
      }
      acc[key] += c[key] || 0;
    }
    return acc;
  }, {});

module.exports = current;
```

### src/_data/seasons/all_competitions.js
```js
const path = require("node:path");
const fg = require("fast-glob");

const DATA_DIR = path.join(__dirname, "*.json");

const current = fg.sync(DATA_DIR)
  .map(filename => require(filename))
  .reduce((acc, curr) => {
    for (const key of Object.keys(curr)) {
      if (!acc[key]) {
        acc[key] = 0;
      }
      acc[key] += curr[key] || 0;
    }
    return acc;
  }, {});

module.exports = current["All Competitions"];
```

### src/index.njk
```njk
---
title: All Time
---

<pre>
{{ seasons.all_time | dump | safe }}
</pre>

<pre>
{{ seasons.all_competitions | dump | safe }}
</pre>
```

### OUTPUT
```html
<pre>
{"All Competitions":42,"League":7}
</pre>

<pre>
42
</pre>
```

**Reply by SteGreig** [reply]
File structure for the data is:
_data/seasons/s2019.json
_data/seasons/s2020.json
_data/seasons/s2021.json
_data/seasons/scurrent.js

The format of all of these data files is the same when outputted directly in an njk template.

_data/all_time.js
```
const requireDir = require('require-dir');
let seasonData = requireDir('./seasons');

seasonData = seasonData.index;

module.exports = seasonData['scurrent'];
```

That outputs the whole data from scurrent.js as expected, but if I try to change that last line to `seasonData['scurrent']['All Club Competitions']` it outputs nothing in the template.

However if I do `seasonData['s2021']['All Club Competitions']` it outputs as expected.

I'm not sure what I'm missing!

**Reply by pdehaan** [reply]
Yeah, this is tricky to debug without seeing the full code and run locally.
I'd probably start by adding a `console.log(seasonData.scurrent)` around your `module.exports` in _data/all_time.js and make sure the data is what you expect and doesn't have extra nesting or is malformed.

I'm also a bit confused by _data/all_time.js and how it differs from just calling _data/seasons/scurrent.js directly, since it seems to mostly be an alias (although I also am not sure what `seasonData = seasonData.index` does so I probably am missing a few things).

**Reply by pdehaan** [reply]
This worked for me, but I'm not sure what `seasonData = seasonData.index` was doing since I didn't have a `seasonIndex.index` property (or presumably _data/seasons/index.js file) and therefore it was setting stuff to `undefined` and throwing errors.

### _data/all_time.js
```js
const requireDir = require('require-dir');
let seasonData = requireDir('./seasons');

console.log("pre", {seasonData});
// seasonData = seasonData.index;

const scurrent = seasonData['scurrent'];
console.log("post", {scurrent});

module.exports = scurrent["All Club Competitions"];
```

### _data/seasons/scurrent.js
Where my _data/seasons/scurrent.js file simply exports out a static value:
```js
module.exports = {
  "All Club Competitions": 99
};
```

### src/index.njk
```njk
---
title: All Time
---

<pre>
{{ all_time | dump | safe }}
</pre>
```

### OUTPUT
```html
<pre>
99
</pre>
```

**Reply by SteGreig** [reply]
I've got rid of the index.js file in /seasons now (thought it was needed but apparently not). Here's a really simplified version of what seasonData returns (the result of doing requireDir on the /seasons directory)...

It's stats for 2 different players for each season.

```
{
     "s2018":{
        "All Club Competitions":[
          [
            {
              "comp":"All Club Competitions",
              "apps":14,
              "goals":1,
              "assists":2
            }
          ],
          [
            {
              "comp":"All Club Competitions",
              "apps":14,
              "goals":1,
              "assists":2
            }
          ]
        ]
     },
     "s2019":{
      "All Club Competitions":[
        [
          {
            "comp":"All Club Competitions",
            "apps":14,
            "goals":1,
            "assists":2
          }
        ],
        [
          {
            "comp":"All Club Competitions",
            "apps":14,
            "goals":1,
            "assists":2
          }
        ]
      ]
    },
    "scurrent":{
      "All Club Competitions":[
        [
          {
            "comp":"All Club Competitions",
            "apps":14,
            "goals":1,
            "assists":2
          }
        ],
        [
          {
            "comp":"All Club Competitions",
            "apps":14,
            "goals":1,
            "assists":2
          }
        ]
      ]
    }
}
```

In all_time.js I want to loop through this object and add all the stats together. However for whatever reason I can't access anything inside of the scurrent part of this object. e.g if I export the following from all_time.js...

`seasonData.s2019` - outputs as expected
`seasonData.scurrent` - outputs as expected

`seasonData.s2019['All Club Competitions']` - outputs as expected
`seasonData.scurrent['All Club Competitions']` - undefined

The latter is required from a .js file (scurrent.js) whereas the rest of the data is from .json files (s2019.json etc) so I figure it must be something to do with that. I was hoping it might be something immediately obvious to someone with more Eleventy experience, but if it's not then don't worry - sure I'll figure it out eventually!

Thanks for your help so far, really appreciate it!

**Reply by pdehaan** [reply]
If I use your static data, it seems to be working for me locally.
### src/all_time.js
```js
const seasonData = {
  "s2018":{
    "All Club Competitions": [
      [ { "comp":"All Club Competitions", "apps":14, "goals":1, "assists":2 } ],
      [ { "comp":"All Club Competitions", "apps":14, "goals":1, "assists":2 } ]
    ]
  },
  "s2019": {
    "All Club Competitions": [
      [ { "comp":"All Club Competitions", "apps":14, "goals":1, "assists":2 } ],
      [ { "comp":"All Club Competitions", "apps":14, "goals":1, "assists":2 } ]
    ]
  },
  "scurrent":{
    "All Club Competitions": [
      [ { "comp":"All Club Competitions", "apps":14, "goals":1, "assists":2 } ],
      [ { "comp":"All Club Competitions", "apps":14, "goals":1, "assists":2 } ]
    ]
  }
};

module.exports = seasonData.scurrent["All Club Competitions"];
```

### src/index.njk
```njk
<pre>
{{ all_time | dump(2) | safe }}
</pre>
```

### OUTPUT
```html
<pre>
[
  [
    {
      "comp": "All Club Competitions",
      "apps": 14,
      "goals": 1,
      "assists": 2
    }
  ],
  [
    {
      "comp": "All Club Competitions",
      "apps": 14,
      "goals": 1,
      "assists": 2
    }
  ]
]
</pre>
```

Yeah, nothing seems immediately obviously wrong looking at the code. But it's tricky to debug without seeing the entire repo and working from the same files. Only other possible thing I can think of is if something is loading async and it's a promise issue, but that seems unlikely because when you were debugging `seasonData.scurrent` it would have probably said it was a pending promise versus an object.

**Reply by SteGreig** [reply]
Ok that's definitely it (a promise issue)... sorry my method of testing was exporting and then outputting in the template.

I'm still not sure how to solve this though, I just can't get my head around promises - nothing I try seems to work. Will have to come back to it with a fresh head tomorrow I think.

Thanks again for your help.
